src/qz/ws/PrintSocketClient.java

```diff
@@ -461,6 +475,106 @@ private void processMessage(Session session, JSONObject json, SocketConnection c
+                } else {
+                    sendError(session, UID, "Already listening to path events");
+                }
+
+                break;
+            }
+            case FILE_STOP_LISTENING: {
+                if (params.isNull("path")) {
+                    connection.removeAllFileListeners();
+                    sendResult(session, UID, null);
+                } else {
+                    Path absPath = FileUtilities.getAbsolutePath(params, shownCertificate, false);
+                    FileIO fileIO = connection.getFileListener(absPath);
+
+                    if (fileIO != null) {
+                        fileIO.close();
+                        FileWatcher.deregisterWatch(fileIO);
+                        connection.removeFileListener(absPath);
+                        sendResult(session, UID, null);
+                    } else {
+                        sendError(session, UID, "Not already listening to path events");
+                    }
+                }
+
+                break;
+            }
+            case FILE_LIST: {
+                Path absPath = FileUtilities.getAbsolutePath(params, shownCertificate, true);
+
+                if (Files.exists(absPath)) {
+                    if (Files.isDirectory(absPath)) {
+                        ArrayList<String> files = new ArrayList<>();
+                        Files.list(absPath).forEach(file -> files.add(file.getFileName().toString()));
+                        sendResult(session, UID, new JSONArray(files));
+                    } else {
+                        log.error("Failed to list '{}' (not a directory)", absPath);
+                        sendError(session, UID, "Path is not a directory");
+                    }
+                } else {
+                    log.error("Failed to list '{}' (does not exist)", absPath);
+                    sendError(session, UID, "Path does not exist");
+                }
+
+                break;
+            }
+            case FILE_READ: {
+                Path absPath = FileUtilities.getAbsolutePath(params, shownCertificate, false);
+                if (Files.exists(absPath)) {
+                    if (Files.isReadable(absPath)) {
+                        sendResult(session, UID, new String(Files.readAllBytes(absPath)));
+                    } else {
+                        log.error("Failed to read '{}' (not readable)", absPath);
+                        sendError(session, UID, "Path is not readable");
+                    }
+                } else {
+                    log.error("Failed to read '{}' (does not exist)", absPath);
+                    sendError(session, UID, "Path does not exist");
+                }
+
+                break;
+            }
+            case FILE_WRITE: {
+                FileParams fileParams = new FileParams(params);
+                Path absPath = FileUtilities.getAbsolutePath(params, shownCertificate, false);
+
+                if (Files.isWritable(absPath)) {
```

Comment:
Some order of operations issue is happening here... this check fails when I try shared: false, sandbox: true and I can't write a file.

Reply:
I'm an idiot. I was trying to remember why I left this check out, thought I tested it with it in, guess not. Files that don't exist can't pass the "isWritable" check.